File: ObjectOrientedPatterns/Behavioral/TemplateMethod.cs
namespace ObjectOrientedPatterns.Behavioral.TemplateMethod
{
    // see https://en.wikipedia.org/wiki/Template_method_pattern

    public abstract class Game
    {
        public void Play()
        {
            Init();

            StartPlay();

            End();
        }

        protected abstract void Init();
        protected abstract void StartPlay();
        protected abstract void End();
    }

    public class Mario : Game
    {
        protected sealed override void Init() => Console.WriteLine("Starting console");
        protected sealed override void StartPlay() => Console.WriteLine("... jump jump jump ...");
        protected sealed override void End() => Console.WriteLine("Thanks for saving me, Mario!");
    }

    public class Chess : Game
    {
        protected sealed override void Init() => Console.WriteLine("Setting up the pieces");
        protected sealed override void StartPlay() => Console.WriteLine("moving pieces around");
        protected sealed override void End() => Console.WriteLine("Checkmate!");
    }
}

File: FunctionalPatterns/Behavioral/TemplateMethod.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Behavioral

module TemplateMethod =

    type Game =
        { init: unit -> unit
          startPlay: unit -> unit
          finish: unit -> unit }

    let playGame game =
        game.init ()
        game.startPlay ()
        game.finish ()

    let mario =
        { init = fun () -> printfn "Starting console"
          startPlay = fun () -> printfn "... jump jump jump ..."
          finish = fun () -> printfn "Thanks for saving me, Mario!" }

    let chess =
        { init = fun () -> printfn "Setting up the pieces"
          startPlay = fun () -> printfn "... moving pieces around ..."
          finish = fun () -> printfn "Checkmate!" }
